executor/common_kvm_arm64.h

```diff
@@ -149,3 +151,95 @@ static volatile long syz_kvm_setup_cpu(volatile long a0, volatile long a1, volat
 
 	return 0;
 }
+#endif
+
+#if SYZ_EXECUTOR || __NR_syz_kvm_vgic_v3_setup
+static int kvm_get_device_attr(int dev_fd, uint32 group, uint64 attr, void* val)
+{
+	struct kvm_device_attr kvmattr = {
+	    .flags = 0,
+	    .group = group,
+	    .attr = attr,
+	    .addr = (uintptr_t)val,
+	};
+
+	return ioctl(dev_fd, KVM_GET_DEVICE_ATTR, &kvmattr);
+}
+
+static int kvm_set_device_attr(int dev_fd, uint32 group, uint64 attr, void* val)
+{
+	struct kvm_device_attr kvmattr = {
+	    .flags = 0,
+	    .group = group,
+	    .attr = attr,
+	    .addr = (uintptr_t)val,
+	};
+
+	return ioctl(dev_fd, KVM_SET_DEVICE_ATTR, &kvmattr);
+}
+
+static int kvm_create_device(int vm_fd, int type)
+{
+	struct kvm_create_device create_dev = {
+	    .type = (uint32)type,
+	    .fd = (uint32)-1,
+	    .flags = 0,
+	};
+
+	if (ioctl(vm_fd, KVM_CREATE_DEVICE, &create_dev) != -1) {
+		return create_dev.fd;
+	} else {
+		return -1;
+	}
+}
+
+#define REDIST_REGION_ATTR_ADDR(count, base, flags, index) \
+	(((uint64)(count) << 52) |                         \
+	 ((uint64)((base) >> 16) << 16) |                  \
+	 ((uint64)(flags) << 12) |                         \
+	 index)
+
+// Set up the VGICv3 interrupt controller.
+// syz_kvm_vgic_v3_setup$arm64(fd fd_kvmvm, ncpus flags[kvm_num_cpus], nirqs flags[kvm_num_irqs])
+static long syz_kvm_vgic_v3_setup(volatile long a0, volatile long a1, volatile long a2)
+{
+	const int vm_fd = a0;
+	const int nr_vcpus = a1;
+	const int want_nr_irq = a2;
+
+	int vgic_fd = kvm_create_device(vm_fd, KVM_DEV_TYPE_ARM_VGIC_V3);
+	if (vgic_fd == -1) {
+		return -1;
+	}
+
+	uint32 nr_irq = 64;
+	kvm_get_device_attr(vgic_fd, KVM_DEV_ARM_VGIC_GRP_NR_IRQS, 0, &nr_irq);
+	nr_irq = want_nr_irq;
+	int ret = kvm_set_device_attr(vgic_fd, KVM_DEV_ARM_VGIC_GRP_NR_IRQS, 0, &nr_irq);
+	if (ret == -1) {
+		close(vgic_fd);
+		return -1;
+	}
+
+	uint64 gicd_base_gpa = ARM64_ADDR_GICD_BASE;
+	ret = kvm_set_device_attr(vgic_fd, KVM_DEV_ARM_VGIC_GRP_ADDR, KVM_VGIC_V3_ADDR_TYPE_DIST, &gicd_base_gpa);
+	if (ret == -1) {
+		close(vgic_fd);
+		return -1;
+	}
+	uint64 redist_attr = REDIST_REGION_ATTR_ADDR(nr_vcpus, ARM64_ADDR_GICR_BASE, 0, 0);
+	ret = kvm_set_device_attr(vgic_fd, KVM_DEV_ARM_VGIC_GRP_ADDR, KVM_VGIC_V3_ADDR_TYPE_REDIST_REGION, &redist_attr);
+	if (ret == -1) {
+		close(vgic_fd);
+		return -1;
+	}
+
+	ret = kvm_set_device_attr(vgic_fd, KVM_DEV_ARM_VGIC_GRP_CTRL, KVM_DEV_ARM_VGIC_CTRL_INIT, NULL);
```

Comment:
It would be useful to describe all these KVM_DEV_ARM_VGIC_GRP_NR_IRQS/KVM_DEV_ARM_VGIC_GRP_ADDR/KVM_DEV_ARM_VGIC_GRP_CTRL attributes in descriptions as well.

syz_kvm_vgic_v3_setup is useful to create a working controlled, but it's all fixed setup, so not very fuzzy.
If we also give all these attributes to the fuzzer, it will be able to create more havoc.